ta/pkcs11/src/object.c

```diff
@@ -0,0 +1,336 @@
+{
+	struct pkcs11_object *obj = NULL;
+
+	obj = TEE_Malloc(sizeof(struct pkcs11_object), TEE_MALLOC_FILL_ZERO);
+	if (!obj)
+		return NULL;
+
+	obj->key_handle = TEE_HANDLE_NULL;
+	obj->attribs_hdl = TEE_HANDLE_NULL;
+	obj->attributes = head;
+
+	return obj;
+}
+
+/*
+ * create_object - create an PKCS11 TA object from its attributes and value
+ *
+ * @session - session requesting object creation
+ * @attributes - reference to serialized attributes
+ * @handle - generated handle for the created object
+ */
+enum pkcs11_rc create_object(void *sess, struct obj_attrs *head,
+			     uint32_t *out_handle)
+{
+	enum pkcs11_rc rc = 0;
+	struct pkcs11_object *obj = NULL;
+	struct pkcs11_session *session = (struct pkcs11_session *)sess;
+	uint32_t obj_handle = 0;
+
+#ifdef DEBUG
+	trace_attributes("[create]", head);
+#endif
+
+	/*
+	 * We do not check the key attributes. At this point, key attributes
+	 * are expected consistent and reliable.
+	 */
+
+	obj = create_obj_instance(head);
+	if (!obj)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	/* Create a handle for the object in the session database */
+	obj_handle = handle_get(&session->object_handle_db, obj);
+	if (!obj_handle) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto err;
+	}
+
+	if (get_bool(obj->attributes, PKCS11_CKA_TOKEN)) {
+		EMSG("Persistent object not yet supported, panic!");
+		TEE_Panic(0);
+	} else {
+		rc = PKCS11_CKR_OK;
+		LIST_INSERT_HEAD(get_session_objects(session), obj, link);
+	}
+
+	*out_handle = obj_handle;
+
+	return PKCS11_CKR_OK;
+err:
+	/* make sure that supplied "head" isn't freed */
+	obj->attributes = NULL;
+	handle_put(&session->object_handle_db, obj_handle);
+	if (get_bool(head, PKCS11_CKA_TOKEN))
+		cleanup_persistent_object(obj, session->token);
+	else
+		cleanup_volatile_obj_ref(obj);
+
+	return rc;
+}
+
+enum pkcs11_rc entry_import_object(struct pkcs11_client *client,
+				   uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct obj_attrs *head = NULL;
+	struct pkcs11_object_head *template = NULL;
+	uint32_t session_handle = 0;
+	size_t template_size = 0;
+	uint32_t obj_handle = 0;
+
+	/*
+	 * Collect the arguments of the request
+	 */
+
+	if (!client || ptypes != exp_pt ||
+	    out->memref.size != sizeof(obj_handle))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get(&ctrlargs, &session_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_attributes(&ctrlargs, &template);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
```

Comment:
Shall `goto out;` since the allocation done by `serialargs_alloc_get_attributes()`.

Ditto at line 231.